compiler/src/formatting/fmt.re

```diff
@@ -0,0 +1,3252 @@
+            } as single_param,
+          ],
+          body,
+        )
+          when label == var =>
+        print_lambda_argument(single_param)
+        ++ string(" =>")
+        ++ print_comment_range(~lead=space, label_loc, body.pexp_loc)
+        ++ group(
+             switch (body.pexp_desc) {
+             | PExpBlock(_) => space ++ print_expression(body)
+             | _ => indent(2, breakable_space ++ print_expression(body))
+             },
+           )
+      | PExpLambda(params, body) =>
+        parens(
+          concat_map(
+            ~lead=
+              next =>
+                print_comment_range(
+                  ~block_start=true,
+                  ~trail=space,
+                  enclosing_start_location(expr.pexp_loc),
+                  next.pla_loc,
+                ),
+            ~sep=
+              (prev, next) =>
+                comma
+                ++ print_comment_range(
+                     ~none=breakable_space,
+                     ~lead=space,
+                     ~trail=breakable_space,
+                     prev.pla_loc,
+                     next.pla_loc,
+                   ),
+            ~trail=
+              last =>
+                trailing_comma
+                ++ print_comment_range(
+                     ~block_end=true,
+                     ~lead=space,
+                     last.pla_loc,
+                     body.pexp_loc,
+                   ),
+            ~f=print_lambda_argument,
+            params,
+          ),
+        )
+        ++ string(" =>")
+        ++ group(
+             switch (body.pexp_desc) {
+             | PExpBlock(_) => space ++ print_expression(body)
+             | _ => indent(2, breakable_space ++ print_expression(body))
+             },
+           )
+      | PExpContinue => string("continue")
+      | PExpBreak => string("break")
+      | PExpTuple(exprs) =>
+        parens(
+          concat_map(
+            ~lead=
+              next =>
+                print_comment_range(
+                  ~block_start=true,
+                  ~trail=space,
+                  enclosing_start_location(expr.pexp_loc),
+                  next.pexp_loc,
+                ),
+            ~sep=
+              (prev, next) =>
+                comma
+                ++ print_comment_range(
+                     ~none=breakable_space,
+                     ~lead=space,
+                     ~trail=breakable_space,
+                     prev.pexp_loc,
+                     next.pexp_loc,
+                   ),
+            ~trail=
+              last =>
+                trailing_comma
+                ++ print_comment_range(
+                     ~block_end=true,
+                     ~lead=space,
+                     last.pexp_loc,
+                     enclosing_end_location(expr.pexp_loc),
+                   ),
+            ~f=print_expression,
+            exprs,
+          ),
+        )
+      | PExpArray([]) => string("[>]")
```

Comment:
I think maybe making `array_brackets` handle this case would be more robust? Also if I'm not mistaken your `array_brackets` implementation would currently break on the `PPatArray([])` since this edge case is not used there

Reply:
It's hard to have `array_brackets` handle this case because it accepts a document as its input, and we don't want to get in the business of inspecting documents and making decisions based on that. I'll add a case for `PPatArray`.

Reply:
I updated `array_brackets` to not include the space, to make adding it the responsibility of the user. I also now ensure comments are printed within an empty list/array.